packages/puppeteer-core/src/bidi/core/Browser.ts

```diff
@@ -0,0 +1,150 @@
+import type * as Bidi from 'chromium-bidi/lib/cjs/protocol/protocol.js';
+
+import {EventEmitter} from '../../common/EventEmitter.js';
+import {throwIfDisposed} from '../../util/decorators.js';
+
+import type {BrowsingContext} from './BrowsingContext.js';
+import type {SharedWorkerRealm} from './Realm.js';
+import type {Session} from './Session.js';
+import {UserContext} from './UserContext.js';
+
+export type AddPreloadScriptOptions = Omit<
+  Bidi.Script.AddPreloadScriptParameters,
+  'functionDeclaration' | 'contexts'
+> & {
+  contexts?: [BrowsingContext, ...BrowsingContext[]];
+};
+
+export class Browser extends EventEmitter<{
+  /** Emitted after the browser closes. */
+  closed: {
+    /** The reason for closing the browser. */
+    reason: string;
+  };
+  /** Emitted after the browser disconnects. */
+  disconnected: {
+    /** The reason for disconnecting the browser. */
+    reason: string;
+  };
+  /** Emitted when a shared worker is created. */
+  sharedworker: {
+    /** The realm of the shared worker. */
+    realm: SharedWorkerRealm;
+  };
+}> {
+  static async from(session: Session): Promise<Browser> {
+    const browser = new Browser(session);
+    await browser.#initialize();
+    return browser;
+  }
+
+  // keep-sorted start
+  #reason: string | undefined;
+  readonly #userContexts = new Map();
+  readonly session: Session;
+  // keep-sorted end
+
+  private constructor(session: Session) {
+    super();
+
+    // keep-sorted start
+    this.session = session;
+    // keep-sorted end
+
+    this.#userContexts.set('', UserContext.create(this, ''));
+  }
+
+  async #initialize() {
+    // ///////////////////////
+    // Connection listeners //
+    // ///////////////////////
+    const connection = this.#connection;
+    connection.on('script.realmCreated', info => {
```

Comment:
I think it would be better if we set up event listeners and send commands through the session rather than exposing the connection to every class?

Reply:
I opted for allowing classes to handle their own events because this provides greater separation. If the session becomes a control plane, then we have to wire the children with the session which typically requires exposing internal state within the children to the session. I.e. this approach keeps the internal state truly internal.

Reply:
I mean that the session can basically `implements Connection` especially given that it is responsible for event subcription

Reply:
I see, so basically add a `send` on the `session`. I've thought about this but thought it's better to separate the implementation because `Session` should precisely match the concept of session in WebDriver BiDi. In particular, FWIS, we don't send the session ID with each command which would in theory make all commands session dependent.

Perhaps though this is the actual intention. If so, the we can do what your saying, but from the spec, I don't see the session being passed anywhere.

Reply:
sessionId is not passed anywhere because there is 1:1 relationship between the session and the connection. In the future, we might support multiple sessions and in that case you always needs to make sure that the connection for the right session object is used (if code deals with multiple sessions/connection). But the main reason for me to do so is that the session and not the connection decides which events to subscribe to. The spec argument is valid but I think it is weak because this is the client code and, for example, the spec session does not set up any event subscriptions whereas the current version does (so technically you need to move out the subscription logic out of the session and leave it to the puppeteer layer). If we move events to session, we could also error if you create an event listener without a subscription.

Reply:
Hmm, you're right about the subscriptions. I forgot to remove that code because we are supposed to subscribe in the puppeteer layer.

I think the multi-session argument makes sense. Let me implement it that way then.

Reply:
so you remove the event subscriptions from session? I think it would be okay to keep them because the library decides which events to listen on on its own so subscription should be also handled in the session, right?